File: Backend/demo/src/main/java/com/example/paintapp/shapes/Freehand.java
package com.example.paintapp.shapes;

import com.example.paintapp.shapes.base.Shape;

import java.awt.Point;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;

public class Freehand extends Shape {

    public Freehand() {
        this.type = "Freehand";
        this.properties.put("points", new ArrayList<Point>());
    }

    @SuppressWarnings("unchecked")
    public void addPoint(int x, int y) {
        List<Point> points = (List<Point>) this.properties.get("points");
        points.add(new Point(x, y));
    }

    @Override
    @SuppressWarnings("unchecked")
    public void move(double dx, double dy) {
        List<Point> points = (List<Point>) this.properties.get("points");
        for (Point p : points) {
            p.x += dx;
            p.y += dy;
        }
    }

    @Override
    public Shape clone() {
        Freehand cloned = (Freehand) super.clone();
        @SuppressWarnings("unchecked")
        List<Point> originalPoints = (List<Point>) this.properties.get("points");
        List<Point> clonedPoints = new ArrayList<>();
        for (Point p : originalPoints) {
            clonedPoints.add(new Point(p.x, p.y));
        }
        cloned.properties.put("points", clonedPoints);
        return cloned;
    }


    //For Debugging
    @Override
    public void draw() {
        @SuppressWarnings("unchecked")
        List<Point> points = (List<Point>) this.properties.get("points");
        System.out.println("Drawing Freehand with points: " + points);
    }
}

File: Frontend/src/app/models/freehand.ts
import { Shape } from "./shape";

export class Freehand extends Shape {
  points: { x: number; y: number }[] = [];

  constructor(x1: number, y1: number, x2: number, y2: number, properties: { [key: string]: any } = {}) {
    super('freehand', x1, y1, x2, y2, properties);
    
    this.points = [{ x: x1, y: y1 }];

    if (!this.properties['points']) {
      this.properties['points'] = this.points;
    }
  }

 
  addPoint(x: number, y: number): void {
    this.points.push({ x, y });
    this.properties['points'] = this.points;
    this.x2 = x;
    this.y2 = y;
  }

  
  getPoints(): { x: number; y: number }[] {
    return this.points;
  }

  
  override toJSON(): any {
    return {
      ...super.toJSON?.() || {},
      type: this.type,
      x1: this.x1,
      y1: this.y1,
      x2: this.x2,
      y2: this.y2,
      points: this.points,
      properties: this.properties
    };
  }
}